ndk/src/utils.rs

```diff
@@ -12,3 +14,71 @@ pub(crate) fn status_to_io_result<T>(status: i32, value: T) -> Result<T> {
         r => unreachable!("Status is positive integer {}", r),
     }
 }
+
+pub(crate) fn android_log(level: Level, tag: &CStr, msg: &CStr) {
+    let prio = match level {
+        Level::Error => ffi::android_LogPriority::ANDROID_LOG_ERROR,
+        Level::Warn => ffi::android_LogPriority::ANDROID_LOG_WARN,
+        Level::Info => ffi::android_LogPriority::ANDROID_LOG_INFO,
+        Level::Debug => ffi::android_LogPriority::ANDROID_LOG_DEBUG,
+        Level::Trace => ffi::android_LogPriority::ANDROID_LOG_VERBOSE,
+    };
+    unsafe {
+        ffi::__android_log_write(prio.0 as c_int, tag.as_ptr(), msg.as_ptr());
+    }
+}
+
+pub(crate) fn log_panic(panic: Box<dyn std::any::Any + Send>) {
+    let rust_panic = unsafe { CStr::from_bytes_with_nul_unchecked(b"RustPanic\0") };
+
+    // Use the Rust logger if installed and enabled, otherwise fall back to the Android system
+    // logger so there is at least some record of the panic
+    let use_log = log_enabled!(Level::Error);
+
+    match panic.downcast::<String>() {
+        Ok(panic_string) => {
+            if use_log {
+                error!("RustPanic: {}", panic_string);
+            } else if let Ok(msg) = CString::new(*panic_string) {
+                android_log(Level::Error, rust_panic, &msg);
+            }
```

Comment:
Wasn't sure about asking this last time, but is there no way to simply abstract away the `if - else if let` to a function in such a way that it can be reused 3 times?

Reply:
I did try, but the third case is dissimilar enough to the first two to make it pretty messy. It's a bit shorter by line count, but total number of characters is about the same. My judgement was that the additional complexity is not worth it to save a few lines.

```rust
pub(crate) fn log_panic(panic: Box<dyn std::any::Any + Send>) {
    fn log_panic_string<S: Into<Vec<u8>> + Display>(tag: &CStr, panic_string: Option<S>) {
        // Use the Rust logger if installed and enabled, otherwise fall back to the Android system
        // logger so there is at least some record of the panic
        if log_enabled!(Level::Error) {
            if let Some(s) = panic_string {
                error!("{}: {}", tag.to_str().unwrap(), s);
            } else {
                error!("{}", tag.to_str().unwrap());
            }
            log::logger().flush();
        } else {
            let panic_c_string = panic_string.map_or(None, |s| CString::new(s).ok());
            let panic_c_str = panic_c_string
                .as_ref()
                .map(|s| s.as_c_str())
                .unwrap_or_default();
            android_log(Level::Error, tag, panic_c_str);
        }
    }

    let rust_panic = unsafe { CStr::from_bytes_with_nul_unchecked(b"RustPanic\0") };
    match panic.downcast::<String>() {
        Ok(panic_string) => log_panic_string(rust_panic, Some(*panic_string)),
        Err(panic) => match panic.downcast::<&str>() {
            Ok(panic_str) => log_panic_string(rust_panic, Some(*panic_str)),
            Err(_) => log_panic_string::<String>(
                unsafe { CStr::from_bytes_with_nul_unchecked(b"UnknownPanic\0") },
                None,
            ),
        },
    }
}
```

Reply:
I would have personally gone for something much more simple. Starting with an `UnknownPanic` still being a `RustPanic`, I prefer keeping the tag the same and getting away of all the complexity around it. Then it becomes as simple as accepting an `Option<&str>`, to which a `Box<String>` also deref's :)

Pushed the suggsetion straight away after "prototyping" this idea, saving you the hassle of having to rewrite it. Let me know what you think!

Reply:
Looks good. The reason I used `<S: Into<Vec<u8>> + Display>` is because it lets `CString::new()` take the inner `Vec` from `String` and push a `0` on the end, potentially saving an allocation. This may be an unnecessary micro-optimisation now that I think of it..

Reply:
It seems likely to me that `length == capacity` in most cases, and a reallocation is necessary either way.

If not we could also implement this with a simple `panic_string: String`, then the owned copy from `&str` to something that contains a `Vec<u8>` happens without generics.

Reply:
I think you're right that `String` is likely to be tightly-packed most of the time. Taking `String` would mean that a `&str` panic requires two allocations (one from `&str` to `String`, then another from `String` to `CString`). I think just leave it as it is with `&str`.